README.md
# freo HTML版（管理画面版）
オリジナルの[freo](http://freo.jp/)はXHTMLなので、これをHTML化しています。<br>[別途カスタマイズしたもの](https://github.com/cccabinet/freo_customize)をベースにしてカスタマイズしています。

:bulb:コメントなどのスパム対策で使用する<strong>[投稿キープラグイン](https://freo.jp/plugin/message.html)と</strong>と<strong>[なぞなぞ認証プラグイン](https://freo.jp/plugin/riddle.html)</strong>はどちらか一方のみに使用を統一してください

### 各リリースバージョンの説明
- [main：フルバージョン](https://github.com/cccabinet/freo_html/tree/main)<br>モバイル用テンプレートを除く全ファイルを収納
- [diff：差替版](https://github.com/cccabinet/freo_html/tree/diff)<br>変更のあったファイルのみを収納
- **admin：管理画面版**（←現在のバージョン）<br>管理用関連ファイルのみを収納


## ファイル構成
★：freoカスタマイズ版から追加
<pre><code>freo_customize/
├── freo/
│   ├── css/
│   │   ├── admin.css（管理画面用）
│   │   ├── common.css（HTML再定義用）
│   │   └── iframe.css（インラインフレーム用）
│   ├── images/
│   │   └── tablesorter/（tablesorter用のファイルを差替）
│   ├── js/
│   │   ├── tinymce/ (TinyMCE用のファイルを追加）
│   │   │   ├── langs/ja.js（日本語表記ファイル）
│   │   │   └── plugins/
│   │   │       ├── freomedia/（freoメディア管理プラグイン）
│   │   │       ├── freomediaform/（freoメディア登録プラグイン）
│   │   │       └── netabare/（ネタバレボタンプラグイン）
│   │   ├── admin.js（管理画面用）
│   │   ├── iframe.js（インラインフレーム用）
│   │   ├── jquery.hidearea.js（ネタバレボタン用を追加）
│   │   └── jQuery.jTagging.min.js（タグの候補をタグクラウド一覧で表示用を追加）
│   ├── libs/
│   │   ├── freo/
│   │   │   ├── internals/
│   │   │   │   ├── admin/
│   │   │   │   │   ├── config.php（管理画面用設定管理ファイル）★
│   │   │   │   │   ├── entry_form.php（管理画面用エントリー入力ファイル）
│   │   │   │   │   ├── media.php（管理画面用メディア管理ファイル）
│   │   │   │   │   ├── media_delete.php（管理画面用メディア削除ファイル）
│   │   │   │   │   └── page_form.php（管理画面用ページ入力ファイル）
│   │   │   │   ├── comment/post.php（コメント登録用ファイル）
│   │   │   │   └── file/default.php（ファイル表示用ファイル）
│   │   │   ├── plugins/
│   │   │   │   ├── begin.entry_convert.php（エントリー本文変換プラグインの開始前ファイル）★
│   │   │   │   ├── begin.information_convert.php（ インフォメーション本文変換プラグインの開始前ファイル）★
│   │   │   │   ├── begin.page_convert.php （ページ本文変換プラグインの開始前ファイル）★
│   │   │   │   ├── config.filemanager.php（ファイル管理プラグインの設定ファイルを追加）
│   │   │   │   ├── display.entry_calender.php（エントリーカレンダー表示プラグインの表示ファイル）
│   │   │   │   └── page.filemanager.php（ファイル管理プラグインのページファイルを追加）
│   │   │   ├── common.php（共通関数ファイル）
│   │   │   ├── initialize.php（初期化処理ファイル）
│   │   │   ├── pictogram.php（絵文字関数ファイル）
│   │   │   ├── transfer.php（セッションID自動付加関数ファイル）
│   │   │   └── version.php（バージョン情報ファイル）
│   │   └── smarty/
│   │       └── plugins/（Smarty用プラグインを追加）
│   └── templates/
│       ├── internals/admin/（管理用htmlテンプレート）
│       └── plugins/
│           ├── export/（エクスポートプラグインのテンプレート）
│           ├── filemanager/（ファイル管理プラグインのテンプレートを追加）
│           └── import/（インポートプラグインのテンプレート）
│
├── markItUp!/
│   ├── freo/
│   │   ├── js/
│   │   │   ├── admin.js（管理画面差替用）
│   │   │   └── iframe.js（インラインフレーム差替用）
│   │   ├── markitup/（markItUp!関連ファイルを差替・追加）
│   │   └── templates/internals/admin/
│   │       ├── entry_form.html（エントリー登録/編集の差替テンプレート）
│   │       ├── footer.html（管理用フッタの差替テンプレートを追加）★
│   │       ├── header.html（管理用ヘッダの差替テンプレート）
│   │       ├── information_form.html（インフォメーション登録/編集の差替テンプレート）
│   │       └── page_form.html（ページ登録/編集の差替テンプレート）
│   └── plugins/
│       └── clap/【拍手送信プラグイン】
│           └── templates/plugins/clap/admin_thank_form.html（お礼登録/編集の差替テンプレート）
│
├── plugins/
│   ├── bookmark/【ブックマーク登録プラグイン】
│   │   └── templates/（管理用テンプレートのみ）
│   ├── bookmark_tagcloud/【ブックマークタグクラウド表示プラグイン】
│   │   └── libs/freo/plugins/config.bookmark_tagcloud.php（設定ファイル）※
│   ├── circle/【サークル管理プラグイン】
│   │   └── templates/（管理用テンプレートのみ）
│   ├── circle_tagcloud/【サークルタグクラウド表示プラグイン】
│   │   └── libs/freo/plugins/config.circle_tagcloud.php（設定ファイル）※
│   ├── clap/【拍手送信プラグイン】
│   │   └── templates/（管理用テンプレートのみ）
│   ├── count/【カウンタプラグイン】
│   │   └── templates/
│   ├── entry_tagcloud/【エントリータグクラウド表示プラグイン】
│   │   └── libs/freo/plugins/config.entry_tagcloud.php（設定ファイル）※
│   ├── entry_tagmanager/【エントリータグ管理プラグイン】
│   │   ├── libs/freo/plugins/
│   │   │   ├── config.entry_tagmanager.php（設定ファイル）
│   │   │   └── end.entry_tagmanager.php（処理終了後表示ファイル）
│   │   └── templates/
│   ├── filter_confirm/【フィルター認証確認プラグイン】
│   │   └── templates/
│   ├── form/【フォーム管理プラグイン】
│   │   └── templates/（管理用テンプレートのみ）
│   ├── menu/【メニュー登録プラグイン】
│   │   └── templates/
│   ├── message/【メッセージ登録プラグイン】
│   │   └── templates/（管理用テンプレートのみ）
│   ├── page_calender/【ページカレンダー表示プラグイン】
│   │   └── libs/freo/plugins/display.page_calender.php（表示ファイル）
│   ├── page_pid_update/【ページ親ID一括変更プラグイン】
│   │   └── templates/
│   ├── page_tagcloud/【ページタグクラウド表示プラグイン】
│   │   └── libs/freo/plugins/config.page_tagcloud.php（設定ファイル）※
│   ├── page_tagmanager/【ページタグ管理プラグイン】
│   │   ├── libs/freo/plugins/
│   │   │   ├── config.page_tagmanager.php（設定ファイル）
│   │   │   └── end.page_tagmanager.php（処理終了後表示ファイル）
│   │   └── templates/
│   ├── paint/【イラスト投稿プラグイン】
│   │   └── templates/（管理用テンプレートのみ）
│   ├── parts/【ブログパーツ管理プラグイン】
│   │   └── templates/
│   ├── popularity/【人気コンテンツプラグイン】
│   │   └── templates/
│   ├── profile/【プロフィール拡張プラグイン】
│   │   └── templates/（管理用テンプレートのみ）
│   ├── riddle/【なぞなぞ認証プラグイン】
│   │   └── templates/
│   └── task/【タスク登録プラグイン】
│       └── templates/
│
└── plugins_catalog/
    ├── catalog/【ショッピングカートプラグイン】
    │   ├── libs/freo/plugins/page.catalog.php（ページファイル）
    │   └── templates/
    │       ├── internals/admin/footer.html（管理用フッタの差替テンプレートを追加）★
    │       └── plugins/catalog/（管理用テンプレートのみ）
    ├── catalog_order/【注文管理プラグイン】
    │   └── templates/
    │       └── plugins/catalog_order/（管理用テンプレート＋cancel_complete.html）
    └── catalog_tagcloud/【ショッピングカートタグクラウド表示プラグイン】
        └── libs/freo/plugins/config.catalog_tagcloud.php（設定ファイル）※
</code></pre>
※のファイルは、[登録/編集画面にタグの候補をタグクラウド一覧で表示する](https://cccabinet.jpn.org/view/69)ために使用

## フォルダの種類
- **freo**<br>freo（v1.21.0）本体（モバイル用テンプレートを除く）のうち、`css/`、`images/`、`js/`、`libs/`、`templates/` の管理用関連ファイルに[ファイル管理プラグイン](https://freo.jp/plugin/filemanager.html)を追加<br>※`freo/css/colorbox.css`、`freo/js/jquery.js`、`freo/js/jquery.colorbox.js`、`freo/js/jquery.tablesorter.js`、`freo/images/colorbox/`、`freo/tinymce/` は不要なので差し替える場合は同ファイルを削除してもOK<br>:bulb:**freo本体を設置したものに上書きして差替えて下さい**
- **markItUp!**<br>[テキストエディタをTinyMCEからmarkItUp!に変更する](https://freo.jp/document/customize/markitup.html)場合のJSやテンプレートの差替ファイル（差替用 `freo/templates/internals/admin/footer.html` を追加）<br>※同梱しているmarkItUp!本体は最新版にし、さらに[拡張版](https://cccabinet.jpn.org/view/102)（一部を除く）を導入済み<br>:bulb:**一旦freo本体の差替版を設置してから上書きして差替えて下さい**（freo/js/tinymce/ は不要なので削除してもOK）
- **plugins**<br>[freoの公式サイトで配布されているプラグイン](https://freo.jp/plugin/)（ファイル管理プラグインとショッピングカート系プラグインを除く）の管理用テンプレートファイルなど<br>:bulb:**プラグイン本体を設置したものに上書きして差替えて下さい**
- **plugins_catalog**<br>[ショッピングカートプラグイン](https://freo.jp/plugin/catalog.html)と[注文管理プラグイン](https://freo.jp/plugin/catalog_order.html)の管理用テンプレートファイル（差替用 `catalog/templates/internals/admin/footer.html` を追加）など<br>:bulb:ショッピングカートプラグインはシェアウェアなので、他のプラグインのテンプレートと分けています<br>:bulb:**プラグイン本体を設置したものに上書きして差替えて下さい**

## カスタマイズの内容
freoのデフォルトのテンプレートは、XHTML1.0で書かれているので、これをHTML仕様に変更して、できるだけ[Google HTML/CSS Style Guide](https://google.github.io/styleguide/htmlcssguide.html)に準ずるようにする。<br>参考：[Google HTML/CSS Style Guide まとめ](https://qiita.com/Sugima/items/785644372397595644ba)
- XHTMLからHTMLへスタイルルールの変更
  - DOCTYPE宣言や `<head>` のメタタグの変更
  - CSSファイルの表記にあるtypeとmediaの部分を削除<br>`<link rel="stylesheet" href="..." type="text/css" media="all" />` ⇒ `<link rel="stylesheet" href="..."  />`
  - JavaScriptファイルの表記にあるtypeの部分を削除<br>`<script type="text/javascript" src="...">` ⇒ `<script src="...">`
  - 空要素が /で綴じられている部分を削除<br>例：`<br />` ⇒ `<br>`
- JSライブラリの設定変更
  - Javascriptファイルの記載を `<head>` ～ `</head>` などから `</body>` の直前に移動（サイトの表示体感速度を高めるため）
  - JavaScriptの実行処理の表記の変更<br>`$(document).ready(function() {}` ⇒ `$(function() {}`（[$(docment).ready(...)は非推奨＋必要ない](https://qiita.com/sukoyakarizumu/items/d07288a3fa67e19b66de)）
- .cssファイルの表記を変更
  - IEはサポートを終了したのでIE6用設定を削除
  - インデントはタブからスペース2つに変更
  - プロパティごとに改行を入れ、それぞれのセレクタと宣言に改行を入れる
  - CSSのプロパティはできる限り一括で指定し、[視覚順](https://qiita.com/mgn/items/6154ccd2e23b2e65c769)に記述
  - できるだけID名とクラス名にタイプセレクタ（要素型セレクタ）の使用を削除（不必要な先祖セレクタを避けることでパフォーマンスを向上させる）<br>例：`div#container` ⇒ `#container`
  - HEX形式のカラーコードは小文字で可能な限り16進数表記の3文字に変更<br>例：`color: #FFFFFF;` ⇒ `color: #fff;`
- コピーライトの表記が `<address>` から `<small>` に変更されたので修正
- 強調タグを `<em>` から `<strong>` に変更（`<em>` は斜字、`<strong>` は太字として使用）<br>TinyMCEとmarkItUp!の設定に斜字(Italic)を追加
- テーブル（表）関連の変更
  - `<table>` の summary 属性が廃止になったので削除
  - HTML5.1から `<tfoot>` を `<tbody>` の前から `</table>` の直前に配置することになったので移動
- フォームのインプット周りのカスタマイズ
  - フォームの項目と入力コントロールをそれぞれ紐付けるため、各項目名に `<label>`、各コントロールまたは `<dd>` に `id` を追加
  - `input type="text"` から各入力欄に合わせたタイプに変更
    - メールアドレス関連：`input type="text"` ⇒ `input type="email"`
    - URL関連：`input type="text"` ⇒ `input type="url"`<br>※メニュー登録プラグインのURL欄は相対パスを入れられるようにするため `input type="text" inputmode="url"` にしています。
    - 電話番号関連：`input type="text"` ⇒ `input type="tel"`
    - キーワード検索関連：`input type="text"` ⇒ `input type="search"`<br>検索項目としてplaceholder属性も追加
    - 数値関連：`input type="text"` ⇒ `input type="text" inputmode="numeric"` または `input type="text" inputmode="decimal"` に変更
  - 一部にplaceholder属性を追加
  - `checked`, `selected` などの属性はXHTMLでは、 `xxx="xxx"` と表記しなければならなかったが、単純に `xxx` に変更
  - 複数アップロードしたい場合は、1つずつ選択欄を増加して行っていたのを、1つの選択欄で複数選択できるようにし、一部のファイルのアップロードを画像のみに制限

:bulb:**[freoカスタマイズ（管理画面版）のREADME.md](https://github.com/cccabinet/freo_customize/tree/admin#readme)のカスタマイズの内容も合わせて参照してください。**

CHANGELOG.md
# 変更ログ

詳しくは、各リリースファイルのREADME.mdを参照してください。

## [1.0.0] - 2022-08-14

### カスタマイズの内容
freoカスタマイズ版（1.0.3）から引き続き以下をカスタマイズ、

- XHTMLからHTMLへスタイルルールの変更
- JSライブラリの設定変更
- .cssファイルの表記を変更
- コピーライトの表記が `<address>` から `<small>` に変更されたので修正
- 強調タグを `<em>` から `<strong>` に変更
- テーブル（表）関連の変更
- フォームのインプット周りのカスタマイズ


freo/libs/freo/prepare.php
<?php

/*********************************************************************

 freo | 設定編集 (2022/08/14)

 Copyright(C) 2009-2022 freo.jp
 customized：cccabinet（https://cccabinet.jpn.org/)

*********************************************************************/

//外部ファイル読み込み
require_once FREO_MAIN_DIR . 'freo/config.php';

//PHP仕様確認
if (isset($_GET['mode']) and $_GET['mode'] == 'checker') {
	require_once FREO_MAIN_DIR . 'freo/checker.php';

	exit;
}

//最低動作環境確認
$errors = array();

if (phpversion() < 5) {
	$errors[] = 'PHP5が利用できません。（現在のバージョンは' . phpversion() . 'です。）';
}
if (!class_exists('pdo')) {
	$errors[] = 'PDOが利用できません。';
}
if (!phpversion('pdo_mysql') and !phpversion('pdo_sqlite')) {
	$errors[] = 'MySQLとSQLiteのいずれも利用できません。';
}
if (!function_exists('mb_language')) {
	$errors[] = 'マルチバイト関数が利用できません。';
}

if (!empty($errors)) {
	$errors[] = '現在の環境ではfreoは動作しません。';

	freo_setup_error(implode('', $errors));
}

//設置URLを反映
if (isset($_POST['token']) and $_POST['token'] == session_id()) {
	//入力内容をチェック
	if ($_POST['url'] == '') {
		freo_setup_error('設置URLが入力されていません。');
	} elseif (strlen($_POST['url']) !== mb_strlen($_POST['url'])) {
		freo_setup_error('設置URLは半角で入力してください。');
	}
	if ($_POST['database_type'] != 'sqlite3' and $_POST['database_type'] != 'sqlite2' and $_POST['database_type'] != 'mysql') {
		freo_setup_error('使用データベースの入力内容が不正です。');
	}
	if ($_POST['database_type'] == 'mysql') {
		if ($_POST['database_host'] != '' and strlen($_POST['database_host']) !== mb_strlen($_POST['database_host'])) {
			freo_setup_error('接続先は半角で入力してください。');
		}
		if ($_POST['database_port'] != '' and strlen($_POST['database_port']) !== mb_strlen($_POST['database_port'])) {
			freo_setup_error('ポートは半角で入力してください。');
		}
		if ($_POST['database_user'] != '' and strlen($_POST['database_user']) !== mb_strlen($_POST['database_user'])) {
			freo_setup_error('ユーザー名は半角で入力してください。');
		}
		if ($_POST['database_password'] != '' and strlen($_POST['database_password']) !== mb_strlen($_POST['database_password'])) {
			freo_setup_error('パスワードは半角で入力してください。');
		}
		if ($_POST['database_charset'] != '' and strlen($_POST['database_charset']) !== mb_strlen($_POST['database_charset'])) {
			freo_setup_error('文字コードは半角で入力してください。');
		}
		if ($_POST['database_name'] != '' and strlen($_POST['database_name']) !== mb_strlen($_POST['database_name'])) {
			freo_setup_error('データベース名は半角で入力してください。');
		}
		if ($_POST['database_prefix'] != '' and strlen($_POST['database_prefix']) !== mb_strlen($_POST['database_prefix'])) {
			freo_setup_error('テーブル名のプレフィックスは半角で入力してください。');
		}
	}

	$_POST['url']               = str_replace("'", "\\'", $_POST['url']);
	$_POST['database_type']     = str_replace("'", "\\'", $_POST['database_type']);
	$_POST['database_host']     = str_replace("'", "\\'", $_POST['database_host']);
	$_POST['database_port']     = str_replace("'", "\\'", $_POST['database_port']);
	$_POST['database_user']     = str_replace("'", "\\'", $_POST['database_user']);
	$_POST['database_password'] = str_replace("'", "\\'", $_POST['database_password']);
	$_POST['database_charset']  = str_replace("'", "\\'", $_POST['database_charset']);
	$_POST['database_name']     = str_replace("'", "\\'", $_POST['database_name']);
	$_POST['database_prefix']   = str_replace("'", "\\'", $_POST['database_prefix']);

	//データベースへ接続テスト
	if ($_POST['database_type'] == 'mysql') {
		try {
			$pdo = new PDO(
				'mysql:dbname=' . $_POST['database_name'] . ';host=' . $_POST['database_host'] . ($_POST['database_port'] ? ';port=' . $_POST['database_port'] : ''),
				$_POST['database_user'],
				$_POST['database_password'],
				array(
					PDO::ATTR_ERRMODE => PDO::ERRMODE_SILENT,
					PDO::MYSQL_ATTR_USE_BUFFERED_QUERY => true
				)
			);
		} catch (PDOException $e) {
			freo_setup_error('データベースへ接続できません : ' . $e->getMessage());
		}
	}

	$url = parse_url($_POST['url']);

	//設定内容を定義
	$configs = array(
		'FREO_HTTP_URL'          => $_POST['url'],
		'FREO_DATABASE_TYPE'     => $_POST['database_type'],
		'FREO_DATABASE_HOST'     => $_POST['database_host'],
		'FREO_DATABASE_PORT'     => $_POST['database_port'],
		'FREO_DATABASE_USER'     => $_POST['database_user'],
		'FREO_DATABASE_PASSWORD' => $_POST['database_password'],
		'FREO_DATABASE_CHARSET'  => $_POST['database_charset'],
		'FREO_DATABASE_NAME'     => $_POST['database_name'],
		'FREO_DATABASE_PREFIX'   => $_POST['database_prefix']
	);

	//chmod使用確認
	if (defined('FREO_PERMISSION_MODE') and FREO_PERMISSION_MODE and @chmod(FREO_TEMPLATE_COMPILE_DIR, 0707)) {
		$chmod = true;
	} else {
		$chmod = false;
	}

	if ($chmod) {
		//パーミッションを変更
		if (@chmod('config.php', FREO_PERMISSION_FILE) === false) {
			freo_setup_error('設定ファイル config.php のパーミッションを変更できません。');
		}

		//データを読み込み
		$data = file_get_contents('config.php');
		if ($data === false) {
			freo_setup_error('設定ファイル config.php を読み込めません。');
		}

		//データを置換
		foreach ($configs as $key => $value) {
			$before = 'define\(\'' . $key . '\', \'.*\'\);';
			$after  = 'define(\'' . $key . '\', \'' . $value . '\');';

			$data = preg_replace('/' . $before . '/', $after, $data);
		}

		//データを書き込み
		if (file_put_contents('config.php', $data) === false) {
			freo_setup_error('設定ファイル config.php に書き込めません。');
		}

		//パーミッションを変更
		if (@chmod('config.php', FREO_PERMISSION_PHP) === false) {
			freo_setup_error('設定ファイル config.php のパーミッションを変更できません。');
		}

		//パーミッションを変更
		if (@chmod(FREO_JS_DIR . 'common.js', FREO_PERMISSION_FILE) === false) {
			freo_setup_error('外部JSファイル ' . FREO_JS_DIR . 'common.js のパーミッションを変更できません。');
		}

		//データを読み込み
		$data = file_get_contents(FREO_JS_DIR . 'common.js');
		if ($data === false) {
			freo_setup_error('外部JSファイル ' . FREO_JS_DIR . 'common.js を読み込めません。');
		}

		//データを置換
		$before = 'var freo_path = \'.*\';';
		$after  = 'var freo_path = \'' . $url['path'] . '\';';

		$data = preg_replace('/' . $before . '/', $after, $data);

		//データを書き込み
		if (file_put_contents(FREO_JS_DIR . 'common.js', $data) === false) {
			freo_setup_error('外部JSファイル ' . FREO_JS_DIR . 'common.js に書き込めません。');
		}

		//パーミッションを変更
		if (@chmod(FREO_JS_DIR . 'common.js', FREO_PERMISSION_JS) === false) {
			freo_setup_error('外部JSファイル ' . FREO_JS_DIR . 'common.js のパーミッションを変更できません。');
		}
	}

	//データ出力
	echo "<!DOCTYPE html>\n";
	echo "<html lang=\"ja\">\n";
	echo "<head>\n";
	echo "<meta charset=\"utf-8\">\n";
	echo "<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n";
	echo "<title>設定編集</title>\n";

	if (file_exists(FREO_CSS_DIR . 'common.css')) {
		echo "<link rel=\"stylesheet\" href=\"" . FREO_CSS_DIR . "common.css\">\n";
	}
	if (file_exists(FREO_CSS_DIR . 'setup.css')) {
		echo "<link rel=\"stylesheet\" href=\"" . FREO_CSS_DIR . "setup.css\">\n";
	}

	echo "</head>\n";
	echo "<body>\n";
	echo "<h1>freo</h1>\n";
	echo "<h2>設定編集</h2>\n";

	if ($chmod) {
		echo "<p>設定ファイルの編集が完了しました。以下のリンクからセットアップに進んでください。</p>\n";
		echo "<p><code>config.php</code> と <code>js/common.js</code> がPHPによって書き換えられているので、必要に応じてダウンロードしておいてください。</p>\n";
		echo "<ul>\n";
		echo "<li><a href=\"index.php/setup\">セットアップ</a></li>\n";
		echo "</ul>\n";
	} else {
		echo "<p>ご利用のサーバーは、<strong>PHPから直接パーミッションを設定することができない</strong>環境のようです。テキストエディタで設定ファイルを編集し、FTPソフトでパーミッションを設定する必要があります。</p>\n";
		echo "<h3>config.phpの編集</h3>\n";
		echo "<p><code>config.php</code> にある設置URLを</p>\n";
		echo "<pre><code>//設置URL\n";
		echo "define('FREO_HTTP_URL', '" . $configs['FREO_HTTP_URL'] . "');</code></pre>\n";
		echo "<p>と設定してください。</p>\n";
		echo "<p>また、データベースの設定を</p>\n";
		echo "<pre><code>//データベースの種類(sqlite3 ... SQLite3 / sqlite2 ... SQLite2 / mysql ... MySQL)\n";
		echo "define('FREO_DATABASE_TYPE', '" . $configs['FREO_DATABASE_TYPE'] . "');\n";
		echo "\n";
		echo "//接続先(MySQL用)\n";
		echo "define('FREO_DATABASE_HOST', '" . $configs['FREO_DATABASE_HOST'] . "');\n";
		echo "\n";
		echo "//ポート(MySQL用)\n";
		echo "define('FREO_DATABASE_PORT', '" . $configs['FREO_DATABASE_PORT'] . "');\n";
		echo "\n";
		echo "//ユーザー名(MySQL用)\n";
		echo "define('FREO_DATABASE_USER', '" . $configs['FREO_DATABASE_USER'] . "');\n";
		echo "\n";
		echo "//パスワード(MySQL用)\n";
		echo "define('FREO_DATABASE_PASSWORD', '" . $configs['FREO_DATABASE_PASSWORD'] . "');\n";
		echo "\n";
		echo "//文字コード(MySQL用)\n";
		echo "define('FREO_DATABASE_CHARSET', '" . $configs['FREO_DATABASE_CHARSET'] . "');\n";
		echo "\n";
		echo "//データベース格納ディレクトリ(SQLite用)\n";
		echo "define('FREO_DATABASE_DIR', 'database/');\n";
		echo "\n";
		echo "//データベース名\n";
		echo "define('FREO_DATABASE_NAME', '" . $configs['FREO_DATABASE_NAME'] . "');\n";
		echo "\n";
		echo "//テーブル名のプレフィックス\n";
		echo "define('FREO_DATABASE_PREFIX', '" . $configs['FREO_DATABASE_PREFIX'] . "');</code></pre>\n";
		echo "<p>と設定してください。（データベースの設定は、もともとの設定内容と同じなら変更の必要はありません。）</p>\n";
		echo "<p>また、セットアップの設定を</p>\n";
		echo "<pre><code>//パーミッション自動設定の利用(true ... 利用する / false ... 利用しない)\n";
		echo "define('FREO_PERMISSION_MODE', false);</code></pre>\n";
		echo "<p>と設定してください。</p>\n";
		echo "<h3>js/common.jsの編集</h3>\n";
		echo "<p><code>js/common.js</code> にある設置パスを</p>\n";
		echo "<pre><code>//設置パス\n";
		echo "var freo_path = '" . $url['path'] . "';</code></pre>\n";
		echo "<p>に設定してください。</p>\n";
		echo "<h3>パーミッションの設定</h3>\n";
		echo "<p>FTPソフトで以下のファイルのパーミッションを設定してください。</p>\n";
		echo "<table>\n";
		echo "<tr><th>対象</th><th>パーミッション</th></tr>\n";
		echo "<tr><td><code>configs/</code> 内のファイル</td><td><code>606</code> に設定</td></tr>\n";
		echo "<tr><td><code>configs/plugins/</code> 内のファイル</td><td><code>606</code> に設定</td></tr>\n";
		echo "<tr><td><code>database/</code></td><td><code>707</code> に設定</td></tr>\n";
		echo "<tr><td><code>database/freo.db</code></td><td><code>606</code> に設定</td></tr>\n";
		echo "<tr><td><code>files/</code> 内のディレクトリ</td><td><code>707</code> に設定</td></tr>\n";
		echo "<tr><td><code>files/temporaries/</code> 内のディレクトリ</td><td><code>707</code> に設定</td></tr>\n";
		echo "<tr><td><code>templates_c/</code></td><td><code>707</code> に設定</td></tr>\n";
		echo "</table>\n";
		echo "<h3>セットアップ</h3>\n";
		echo "<p>設定ファイルの編集とパーミッションの設定が完了したら、以下のリンクからセットアップに進んでください。</p>\n";
		echo "<ul>\n";
		echo "<li><a href=\"index.php/setup\">セットアップ</a></li>\n";
		echo "</ul>\n";
	}

	echo "</body>\n";
	echo "</html>\n";

	exit;
}

//設置URLを取得
$url = 'http://' . $_SERVER['SERVER_NAME'] . $_SERVER['SCRIPT_NAME'];

if (preg_match('/^(.+)index.php$/', $url, $matches)) {
	$url = $matches[1];
}

//データ出力
echo "<!DOCTYPE html>\n";
echo "<html lang=\"ja\">\n";
echo "<head>\n";
echo "<meta charset=\"utf-8\">\n";
echo "<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n";
echo "<title>設定編集</title>\n";

if (file_exists(FREO_CSS_DIR . 'common.css')) {
	echo "<link rel=\"stylesheet\" href=\"" . FREO_CSS_DIR . "common.css\">\n";
}
if (file_exists(FREO_CSS_DIR . 'setup.css')) {
	echo "<link rel=\"stylesheet\" href=\"" . FREO_CSS_DIR . "setup.css\">\n";
}
echo "</head>\n";
echo "<body>\n";
echo "<h1>freo</h1>\n";
echo "<h2>設定編集</h2>\n";

echo "<ul>\n";
echo "<li><a href=\"index.php?mode=checker\">PHPの仕様を確認する</a>。</li>\n";
echo "</ul>\n";
echo "<ul>\n";
echo "<li>freoの設定ファイルを編集します。</li>\n";
echo "<li>各情報を入力してください。</li>\n";
echo "</ul>\n";
echo "<form action=\"index.php\" method=\"post\">\n";
echo "<fieldset>\n";
echo "<legend>設定ファイル編集実行フォーム</legend>\n";
echo "<input type=\"hidden\" name=\"token\" value=\"" . session_id() . "\">\n";

echo "<dl>\n";
echo "<dt><label for=\"url\">設置URL</label></dt>\n";
echo "<dd><input type=\"url\" name=\"url\" id=\"url\" size=\"30\" value=\"" . $url . "\"></dd>\n";
echo "<dt><label for=\"database_type\">使用データベース</label></dt>\n";
echo "<dd>\n";
echo "<select name=\"database_type\" id=\"database_type\">\n";

if (phpversion('pdo_sqlite')) {
	echo "<option value=\"sqlite3\">SQLite3</option>\n";
	echo "<option value=\"sqlite2\">SQLite2</option>\n";
}
if (phpversion('pdo_mysql')) {
	echo "<option value=\"mysql\">MySQL</option>\n";
}

echo "</select>\n";
echo "</dd>\n";
echo "</dl>\n";

echo "<dl id=\"database_mysql\">\n";
echo "<dt><label for=\"host\">接続先</label></dt>\n";
echo "<dd><input type=\"text\" name=\"database_host\" id=\"host\" size=\"30\" value=\"\"></dd>\n";
echo "<dt><label for=\"port\">ポート</label></dt>\n";
echo "<dd><input type=\"text\" name=\"database_port\" id=\"port\" size=\"30\" value=\"\"></dd>\n";
echo "<dt><label for=\"user\">ユーザー名</label></dt>\n";
echo "<dd><input type=\"text\" name=\"database_user\" id=\"user\" size=\"30\" value=\"\"></dd>\n";
echo "<dt><label for=\"password\">パスワード</label></dt>\n";
echo "<dd><input type=\"text\" name=\"database_password\" id=\"password\" size=\"30\" value=\"\"></dd>\n";
echo "<dt><label for=\"charset\">文字コード</label></dt>\n";
echo "<dd><input type=\"text\" name=\"database_charset\" id=\"charset\" size=\"30\" value=\"\"></dd>\n";
echo "</dl>\n";

echo "<dl>\n";
echo "<dt><label for=\"database_name\">データベース名</label></dt>\n";
echo "<dd><input type=\"text\" name=\"database_name\" id=\"database_name\" size=\"30\" value=\"freo.db\"></dd>\n";
echo "<dt><label for=\"prefix\">テーブル名のプレフィックス</label></dt>\n";
echo "<dd><input type=\"text\" name=\"database_prefix\" id=\"prefix\" size=\"30\" value=\"freo_\"></dd>\n";
echo "</dl>\n";

echo "<p><input type=\"submit\" value=\"設定編集\"></p>\n";
echo "</fieldset>\n";
echo "</form>\n";

if (file_exists(FREO_JS_DIR . 'jquery.js')) {
	echo "<script src=\"https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js\" integrity=\"sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=\" crossorigin=\"anonymous\"></script>\n";
	echo "<script>\n";
	echo "$(function() {\n";
	echo "	if ($('#database_type').val() != 'mysql') {\n";
	echo "		$('#database_mysql').hide();\n";
	echo "	}\n";
	echo "	$('#database_type').change(function() {\n";
	echo "		if ($(this).val() == 'mysql') {\n";
	echo "			$('#database_mysql').show();\n";
	echo "		} else {\n";
	echo "			$('#database_mysql').hide();\n";
	echo "		}\n";
	echo "	});\n";
	echo "});\n";
	echo "</script>\n";
}

echo "</body>\n";
echo "</html>\n";

exit;

/* エラー表示 */
function freo_setup_error($message)
{
	if (is_array($message)) {
		list($pdo_state, $pdo_code, $pdo_message) = $message;

		$message = $pdo_message;
	}

	echo "<!DOCTYPE html>\n";
	echo "<html lang=\"ja\">\n";
	echo "<head>\n";
	echo "<meta charset=\"utf-8\">\n";
	echo "<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n";
	echo "<title>エラーが発生しました</title>\n";

	if (file_exists(FREO_CSS_DIR . 'common.css')) {
		echo "<link rel=\"stylesheet\" href=\"" . FREO_CSS_DIR . "common.css\">\n";
	}

	echo "</head>\n";
	echo "<body>\n";
	echo "<h1>freo</h1>\n";
	echo "<h2>エラーが発生しました</h2>\n";
	echo "<p>" . $message . "</p>\n";
	echo "</body>\n";
	echo "</html>\n";

	exit;
}

?>
